src/main/java/frc/robot/vision/sources/LimeLightSource.java

```diff
@@ -0,0 +1,153 @@
+				Alert.AlertType.WARNING,
+				this.getLogPath() + "DisconnectedAt",
+				() -> getLimelightNetworkTableEntry("tv").getInteger(-1) == -1
+			)
+		);
+	}
+
+	public void switchToOldBotPose(boolean useOldRobotPose) {
+		useOldRobotPoseEntry = useOldRobotPose;
+	}
+
+	@Override
+	public void updateGyroAngles(GyroAngleValues gyroAngleValues) {
+		this.gyroAngleValues = gyroAngleValues;
+	}
+
+	@Override
+	public void updateEstimation() {
+		robotOrientationEntry.setDoubleArray(
+			new double[] {
+				gyroAngleValues.yaw().getDegrees(),
+				gyroAngleValues.yawRate(),
+				gyroAngleValues.pitch().getDegrees(),
+				gyroAngleValues.pitchRate(),
+				gyroAngleValues.roll().getDegrees(),
+				gyroAngleValues.rollRate()}
+		);
+		robotPoseArray = (useOldRobotPoseEntry ? oldRobotPoseEntry : robotPoseEntry)
+			.getDoubleArray(new double[VisionConstants.LIMELIGHT_ENTRY_ARRAY_LENGTH]);
+		aprilTagPoseArray = aprilTagPoseEntry.getDoubleArray(new double[VisionConstants.LIMELIGHT_ENTRY_ARRAY_LENGTH]);
+		double[] robotPoseWithoutGyroInput = oldRobotPoseEntry.getDoubleArray(new double[VisionConstants.LIMELIGHT_ENTRY_ARRAY_LENGTH]);
+		robotHeading = Rotation2d.fromDegrees(robotPoseWithoutGyroInput[LimelightEntryValue.YAW_ANGLE.getIndex()]);
+	}
+
+	private Optional<Pair<Pose3d, Double>> getUpdatedPose3DEstimation() {
+		int id = (int) aprilTagIdEntry.getInteger(VisionConstants.NO_APRILTAG_ID);
+		if (id == VisionConstants.NO_APRILTAG_ID) {
+			return Optional.empty();
+		}
+
+		double processingLatencySeconds = Conversions.milliSecondsToSeconds(robotPoseArray[LimelightEntryValue.TOTAL_LATENCY.getIndex()]);
+		double timestamp = TimeUtils.getCurrentTimeSeconds() - processingLatencySeconds;
+
+		Pose3d robotPose = new Pose3d(
+			getPoseValue(LimelightEntryValue.X_AXIS),
+			getPoseValue(LimelightEntryValue.Y_AXIS),
+			getPoseValue(LimelightEntryValue.Z_AXIS),
+			new Rotation3d(
+				Math.toRadians(getPoseValue(LimelightEntryValue.ROLL_ANGLE)),
+				Math.toRadians(getPoseValue(LimelightEntryValue.PITCH_ANGLE)),
+				Math.toRadians(getPoseValue(LimelightEntryValue.YAW_ANGLE))
+			)
+		);
+		return Optional.of(new Pair<>(robotPose, timestamp));
+	}
+
+	public double getPoseValue(LimelightEntryValue entryValue) {
+		return robotPoseArray[entryValue.getIndex()];
+	}
+
+	public double getAprilTagValue(LimelightEntryValue entryValue) {
+		return aprilTagPoseArray[entryValue.getIndex()];
+	}
+
+	@Override
+	public Optional<RawVisionAprilTagData> getRawVisionEstimation() {
+		Optional<Pair<Pose3d, Double>> poseEstimation = getUpdatedPose3DEstimation();
+		return poseEstimation.map(
+			pose3dDoublePair -> new RawVisionAprilTagData(
+				pose3dDoublePair.getFirst(),
+				getAprilTagValue(LimelightEntryValue.Y_AXIS),
+				getAprilTagValue(LimelightEntryValue.Z_AXIS),
+				pose3dDoublePair.getSecond()
+			)
+		);
+	}
+
+	@Override
+	public boolean shallBeFiltered() {
+		return getRawVisionEstimation().map(filter::doesFilterPasses).orElseGet(() -> false);
```

Comment:
what? why?

Reply:
??? -60 points

Reply:
What what 

We said that the filters are in the sources now

Reply:
You mean that I should've map it to true?